pravega/pravega
# Issue 6139: (SegmentStore) Improving Table Segment Pre-Caching and Background Indexing

**Change log description**  
- Made `TableExtensionConfig` externally-configurable. This should allow us to properly tweak this if we have to (for debugging/repair purposes).
- Changed `ContainerKeyIndex` pre-caching to execute in batches, no bigger than 128MB. This will enable precaching of larger segments without the risk of running out of memory.
- Changed `WriterTableProcessor` to index a maximum of 128MB at once, even if the backlog is bigger. Remaining unindexed data will be processed in later iterations.

**Purpose of the change**  
Fixes #6139.

**What the code does**  
- `TableExtensionConfig` 
    - This defines the major config points for the `ContainerTableExtensionConfig` component. The original purpose was to make unit testing easier.
    - This change converts it to an externally-configurable component (just like any other config components) which should enable us to tweak these values if we need to get the Pravega cluster out of a bad state.
- `ContainerKeyIndex` pre-caching
    - Using an async loop to break down the pre-cache segment range into smaller ranges and process them in sequence. This way, no more than 128MB (or whatever is configured) of data will be read in memory at any given time.
 - `WriterTableProcessor`
     - Changed the `flush` behavior to cap the amount of data to process at 128MB. The actual exact amount of data will vary, as it will try to stop on an append boundary (otherwise it may try to stop in the middle of a table entry).


**How to verify it**  
All existing unit tests must pass - this was an "idempotent" change.
Some unit tests have been updated to check for new cases and new unit tests have been added to verify the same.
